soc: nordic: nrf53: Forward gpio pins to network core just before reset

nordic-krch opened the pull request:
Instead of forwarding pins to the network core during the initialization, do it just before reseting the network core. With this approach pins can be used by the application core as long as network core is not started.

joerchan reviewed (commented):
I guess it doesn't hurt, this one is usually just selected from BOARD_ENABLE_CPUNET though.

nordic-krch commented:
I hit that issue when I was trying to use SWO on cpuapp-only app. SWO is using same pin that is defined in the gpio forwarder on nrf5340dk (used for net UART). Pin was assigned to cpunet even though cpunet was not used at all.

anangl commented on soc/nordic/nrf53/Kconfig:
```diff
@@ -165,6 +165,7 @@ config NRF_SPU_RAM_REGION_SIZE
 config SOC_NRF_GPIO_FORWARDER_FOR_NRF5340
 	bool "Forward GPIO pins to network core"
 	depends on NRF_SOC_SECURE_SUPPORTED
+	depends on SOC_NRF53_CPUNET_ENABLE
```
I don't think this is correct. One may still want to forward pins to the Network core without enabling it on boot (this can be done later with a call to `nrf53_cpunet_enable()`).

nordic-krch replied:
is there any Kconfig that enables that API or is it always available?

anangl commented:
> I hit that issue when I was trying to use SWO on cpuapp-only app. SWO is using same pin that is defined in the gpio forwarder on nrf5340dk (used for net UART). Pin was assigned to cpunet even though cpunet was not used at all.

Why not just use `CONFIG_SOC_NRF_GPIO_FORWARDER_FOR_NRF5340=n` in your configuration?

nordic-krch commented:
> Why not just use CONFIG_SOC_NRF_GPIO_FORWARDER_FOR_NRF5340=n in your configuration (or disable the forwarder node)?

You need to be aware of that. When you enable SWO then system init function from MDK configures SWO pin CTRLSEL to Trace&Debug then gpio forwarder overwrites it and SWO does not work. I had to spent some time to figure that pins forwarded to cpunet even though i'm not using it.

anangl commented:
> Maybe gpios shall be forwarded to cpunet when cpunet is taken out of reset here?
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/fb43f2d0ecb3b93fbd31b131cb5b40d378cc38da/soc/nordic/nrf53/nrf53_cpunet_mgmt.c#L26

Yes, that seems to be reasonable.

nordic-krch committed:
soc: nordic: nrf53: Forward gpio pins to network core just before reset

Instead of forwarding pins to the network core during the
initialization, do it just before reseting the network core. With
this approach pins can be used by the application core as long as
network core is not started.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
I've moved forwarding pins to the function that resets the network core so until network core is used pins can be used by the app core.

anangl commented on soc/nordic/nrf53/nrf53_cpunet_mgmt.c:
```diff
@@ -32,6 +55,11 @@ static void onoff_stop(struct onoff_manager *mgr, onoff_notify_fn notify)
 
+	if (IS_ENABLED(CONFIG_SOC_NRF_GPIO_FORWARDER_FOR_NRF5340)) {
+		/* Assign pins back to app core. */
+		forward_pins(NRF_GPIO_PIN_SEL_APP);
```
nit: `forward_pins` looks a bit weird in this context, perhaps the function should be renamed to e.g. `assign_pins`

nordic-krch replied:
done

anangl commented on soc/nordic/nrf53/nrf53_cpunet_mgmt.c:
```diff
@@ -21,8 +21,31 @@
+	};
+
+	for (int i = 0; i < ARRAY_SIZE(forwarded_psels); i++) {
+		soc_secure_gpio_pin_mcu_select(forwarded_psels[i], sel);
```
Don't you need to include `<soc_secure.h>` for this?

nordic-krch commented:
@anangl @kl-cruz  can you take a look? Had to add ifdef for non-secure build.

joerchan commented on soc/nordic/nrf53/nrf53_cpunet_mgmt.c:
```diff
@@ -15,14 +15,40 @@
+
+static void assign_pins(nrf_gpio_pin_sel_t sel)
+{
+#if !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
```
Now this is incorrect.

I suggest `ifdef CONFIG_SOC_NRF_GPIO_FORWARDER_FOR_NRF5340` as it was, instead of relying on other configs like it was.

SOC_NRF_GPIO_FORWARDER_FOR_NRF5340 depends on NRF_SOC_SECURE_SUPPORTED,
NRF_SOC_SECURE_SUPPORTED depends on  !TRUSTED_EXECUTION_NONSECURE || (BUILD_WITH_TFM && TFM_PARTITION_PLATFORM).

Don't duplicate this logic, you can do the forwarding if TF-M is enabled with the platform partition enabled.
